File: go/dsa/algorithms/sliding-window/permutation.go
func checkInclusion(target string, searchString string) bool {
    if len(target) > len(searchString) { return false }

    targetFreq := make([]int, 26)
    windowFreq := make([]int, 26)
    for i := 0; i < len(target); i++ {
        targetFreq[target[i]-'a']++
    }

    left := 0
    for right := 0; right < len(searchString); right++ {
        // 1️⃣ Expand
        windowFreq[searchString[right]-'a']++

        // 2️⃣ When window size reaches k
        if right - left + 1 == len(target) {
            // 3️⃣ Record / Evaluate
            if checkMatch(targetFreq, windowFreq) {
                return true
            }

            // 4️⃣ Slide window
            windowFreq[searchString[left]-'a']--
            left++
        }
    }
    return false
}

func checkMatch(targetFreq []int, windowFreq []int) bool {
    for j := 0; j < 26; j++ {
        if targetFreq[j] != windowFreq[j] {
            return false
        }
    }
    return true
}

File: javascript/dsa/algorithms/sliding-window/permutation.js
function checkInclusion(target, searchString) {
    if (target.length > searchString.length) return false;

    let targetFreq = new Array(26).fill(0);
    let windowFreq = new Array(26).fill(0);
    for (let i = 0; i < target.length; i++) {
        targetFreq[target.charCodeAt(i) - 97]++;
    }

    let left = 0;
    for (let right = 0; right < searchString.length; right++) {
        // 1️⃣ Expand
        windowFreq[searchString.charCodeAt(right) - 97]++;

        // 2️⃣ When window size reaches k
        if (right - left + 1 === target.length) {
            // 3️⃣ Record / Evaluate
            if (checkMatch(targetFreq, windowFreq)) {
                return true;
            }

            // 4️⃣ Slide window
            windowFreq[searchString.charCodeAt(left) - 97]--;
            left++;
        }
    }
    return false;
}

function checkMatch(targetFreq, windowFreq) {
    for (let j = 0; j < 26; j++) {
        if (targetFreq[j] !== windowFreq[j]) {
            return false;
        }
    }
    return true;
}